rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala

```diff
@@ -0,0 +1,193 @@
+  DeleteJoins,
+  HotStoreAction,
+  InsertContinuations,
+  InsertData,
+  InsertJoins,
+  Serialize
+}
+import java.nio.charset.StandardCharsets
+import coop.rchain.rspace.internal._
+import scodec.Codec
+import scodec.bits.{BitVector, ByteVector}
+import coop.rchain.rspace.nextgenrspace.history.HistoryRepositoryImpl._
+
+object HistoryTransformers {
+
+  // TODO more readable if case class
+  type Result = (Blake2b256Hash, Option[PersistedData], HistoryAction)
+
+  final case class MeasureJournal(
+      key: String,
+      size: Long,
+      action: String,
+      datumSize: Int,
+      datumLen: Long,
+      continuationsLength: Long,
+      continuationsSize: Int
+  )
+
+  // TODO more readable if case class
+  type JournalResultTuple = (MeasureJournal, Result)
+
+  def create[C, P, A, K]()(
+      implicit
+      codecC: Codec[C],
+      codecP: Codec[P],
+      codecA: Codec[A],
+      codecK: Codec[K]
+  ): HistoryTransformers[C, P, A, K] =
+    new HistoryTransformers[C, P, A, K]()(codecC, codecP, codecA, codecK): HistoryTransformers[
+      C,
+      P,
+      A,
+      K
+    ]
+}
+
+@SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+class HistoryTransformers[C, P, A, K](
+    implicit
+    codecC: Codec[C],
+    codecP: Codec[P],
+    codecA: Codec[A],
+    codecK: Codec[K]
+) {
+
+  import HistoryTransformers._
+
+  implicit val serializeC: Serialize[C] = Serialize.fromCodec(codecC)
+  private val dataSuffixBits            = BitVector("-data".getBytes(StandardCharsets.UTF_8))
+  private val continuationSuffixBits    = BitVector("-continuation".getBytes(StandardCharsets.UTF_8))
+  private val joinSuffixBits            = BitVector("-joins".getBytes(StandardCharsets.UTF_8))
+
+  private def hashWithSuffix(bits: BitVector, suffix: BitVector): Blake2b256Hash = {
+    val suffixed = bits ++ suffix
+    Blake2b256Hash.create(suffixed.toByteVector)
+  }
+
+  def hashJoinsChannel(channel: C): Blake2b256Hash =
+    hashWithSuffix(codecC.encode(channel).get, joinSuffixBits)
+
+  def hashContinuationsChannels(channels: Seq[C]): Blake2b256Hash = {
+    val chs = channels
+      .map(c => serializeC.encode(c))
+      .sorted(util.ordByteVector)
+    val channelsBits = codecSeq(codecByteVector).encode(chs).get
+    hashWithSuffix(channelsBits, continuationSuffixBits)
+  }
+
+  def hashDataChannel(channel: C): Blake2b256Hash =
+    hashWithSuffix(codecC.encode(channel).get, dataSuffixBits)
+
+  def decode[R](bv: ByteVector)(implicit codecR: Codec[R]): Seq[R] =
+    Codec.decode[Seq[R]](bv.bits).get.value
+
+  def transform(actions: List[HotStoreAction]): List[JournalResultTuple] =
+    actions.map {
+      case i: InsertData[C, A] =>
+        val key      = hashDataChannel(i.channel)
+        val data     = encodeData(i.data)
+        val dataLeaf = DataLeaf(data)
+        val dataHash = Blake2b256Hash.create(data)
+        (
+          MeasureJournal(
+            key = key.bytes.toHex,
+            size = data.toBitVector.size,
```

Comment:
this is a bug from the original design - bits needs to be divided by 8 or `toByteVector.size` needs to be used.